presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -125,43 +128,45 @@ public SemiTransactionalHiveMetastore(
             HdfsEnvironment hdfsEnvironment,
             ExtendedHiveMetastore delegate,
             ListeningExecutorService renameExecutor,
+            HiveMetastoreAuthentication metastoreAuthentication,
             boolean skipDeletionForAlter,
             boolean skipTargetCleanupOnRollback)
     {
         this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
         this.delegate = requireNonNull(delegate, "delegate is null");
         this.renameExecutor = requireNonNull(renameExecutor, "renameExecutor is null");
+        this.metastoreAuthentication = requireNonNull(metastoreAuthentication, "metastoreAuthentication is null");
         this.skipDeletionForAlter = skipDeletionForAlter;
         this.skipTargetCleanupOnRollback = skipTargetCleanupOnRollback;
     }
 
-    public synchronized List<String> getAllDatabases()
+    public synchronized List<String> getAllDatabases(ConnectorIdentity identity)
     {
         checkReadable();
-        return delegate.getAllDatabases();
+        return metastoreAuthentication.doAs(identity.getUser(), delegate::getAllDatabases);
     }
 
-    public synchronized Optional<Database> getDatabase(String databaseName)
+    public synchronized Optional<Database> getDatabase(ConnectorIdentity identity, String databaseName)
     {
         checkReadable();
-        return delegate.getDatabase(databaseName);
+        return metastoreAuthentication.doAs(identity.getUser(), () -> delegate.getDatabase(databaseName));
     }
 
-    public synchronized Optional<List<String>> getAllTables(String databaseName)
+    public synchronized Optional<List<String>> getAllTables(ConnectorIdentity identity, String databaseName)
     {
         checkReadable();
         if (!tableActions.isEmpty()) {
             throw new UnsupportedOperationException("Listing all tables after adding/dropping/altering tables/views in a transaction is not supported");
         }
-        return delegate.getAllTables(databaseName);
+        return metastoreAuthentication.doAs(identity.getUser(), () -> delegate.getAllTables(databaseName));
     }
 
-    public synchronized Optional<Table> getTable(String databaseName, String tableName)
+    public synchronized Optional<Table> getTable(ConnectorIdentity identity, String databaseName, String tableName)
     {
         checkReadable();
         Action<TableAndMore> tableAction = tableActions.get(new SchemaTableName(databaseName, tableName));
         if (tableAction == null) {
-            return delegate.getTable(databaseName, tableName);
+            return metastoreAuthentication.doAs(identity.getUser(), () -> delegate.getTable(databaseName, tableName));
```

Comment:
This is still too high level API to wrap in the `doAs` statement. As explained in the comment [here](https://github.com/prestodb/presto/pull/14464#issuecomment-642826246) different metastore implementations (Hive / Glue / Inmemory / Thrift / File / etc.) have completely different authentication methods. The `doAs` mechanism is HiveMetastore specific (even Hadoop client specific). Using it at this level breaks the encapsulation and abstraction.

Additionally, historically in Presto we are trying to minimize the usage of the thread local contexts, as they tend to be very fragile. Think if there's a metastore implementation that does method invokation in a separate thread pool for efficiency (e.g.: multiple calls in parallel). The `doAs` based implementation will be broken for such a use case.

Thus I strongly recommend you passing the identity all the way down to the lower level interface and wrapping the invokations with `doAs `only at the `ThriftHiveMetastore.class` level

Reply:
Ping again, the doAs has been removed, and the new implementation is based on passing the MetastoreContext object explicitly in the interface.